private/bufpkg/bufmodule/bufmoduleapi/uploader.go

```diff
@@ -111,38 +111,17 @@ func (a *uploader) Upload(
 	if err != nil {
 		return nil, err
 	}
-
-	// This must be in the same order as contentModules.
-	var modules []*modulev1.Module
-	if uploadOptions.CreateIfNotExist() {
-		// We must attempt to create each module one at a time, since CreateModules will return
-		// an `AlreadyExists` if any of the modules we are attempting to create already exists,
-		// and no new modules will be created.
-		modules = make([]*modulev1.Module, len(contentModules))
-		for i, contentModule := range contentModules {
-			module, err := a.createContentModuleIfNotExist(
-				ctx,
-				primaryRegistry,
-				contentModule,
-				uploadOptions.CreateModuleVisibility(),
-				uploadOptions.CreateDefaultLabel(),
-			)
-			if err != nil {
-				return nil, err
-			}
-			modules[i] = module
-		}
-	} else {
-		// The modules retrieved by GetModules retains the same order as the request, so
-		// this matches the order of contentModules.
-		modules, err = a.validateContentModulesExist(
-			ctx,
-			primaryRegistry,
-			contentModules,
-		)
-		if err != nil {
-			return nil, err
-		}
+	// We must validate that the modules exist, and if they do not, create them.
```

Comment:
This is a tad confusing, as the RPC will return `NotFound` if *any* of the N modules does not exist. If so, the create logic is called. Is this what we want? If so, we should update the comments to reflect this is what is going on and why this is OK

Reply:
Yes I believe that is what we want. Get all the modules, which will succeed for the happy path where all modules already exist. If some do not exist a 404 error is returned. We then iterate over each module trying to create it as it's not clear which modules failed the fetch. Will update the comments.

Reply:
I commented on the issue in detail, but just responding with my 2c here: our API returns an error that explicitly and clearly handles the case where a module already exists for `CreateModules`, so it feels counter-intuitive for a client that intends to call `CreateModules` to also call `GetModules` beforehand to validate.

Reply:
We can probably close this PR, @doriable brings up some valid points. We chatted on the issue and the consensus was to change the behavior elsewhere, instead of in the CLI code.